Added initial logIn screen and functionality

alexbtlv opened the pull request:
Added logIn screen with three pages for a tutorial. Of-course we can add or remove as many pages as we want. All pages are absolutely customizable. I used [BWWalkthrough](https://github.com/ariok/BWWalkthrough).

Added Parse framework and a user authorization via Twitter.

##IMPORTANT
You will have to create Keys.plist file locally on your machine and be sure to add it to your .gitignore file to prevent sharing keys publicly. Check out: [How to hide api keys in GitHub for iOS swift projects](http://stackoverflow.com/questions/30803244/how-to-hide-api-keys-in-github-for-ios-swift-projects)
Add the following keys to the file:
1. parseApplicationId
2. parseClientKey
3. twitterConsumerKey
4. twitterConsumerSecret

You should be able to find Parse keys in your Parse / Application / Settings / Keys page. 
However Twitter does not allow developers to share their applications, so DM me on Slack (@AlexBtlv) and will send you the twitter keys. Please be careful with them.

##TODO
1. If we're going to support both iPhone orientations there is an issue when user changes device orientation while on logIn screen

IanKeen commented on Reading List App/Reading List App/AppDelegate.swift:
```diff
@@ -16,6 +18,33 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
+        if let path = NSBundle.mainBundle().pathForResource("Keys", ofType: "plist") {
+            keys = NSDictionary(contentsOfFile: path)
+        }
+        if let dict = keys {
```
Do we want to promote 'shadowing' ? i.e. 
`if let dict = dict {` 
I think this reduces the complexity for readers

IanKeen commented on Reading List App/Reading List App/BWWalkthrough/BWWalkthroughViewController.swift:
```diff
@@ -0,0 +1,285 @@
+
+
+@objc class BWWalkthroughViewController: UIViewController, UIScrollViewDelegate {
+    
```
Is there a reason theres so much ui/autolayout code in the file? can it be done in interface builder to reduce the amount of code?

fcanas replied:
Looks like this is an external dependency that happens to have been included in the PR directly?

IanKeen commented on Reading List App/Reading List App/FirstViewController.swift:
```diff
@@ -7,19 +7,61 @@
+        }
     }
 
     override func didReceiveMemoryWarning() {
```
remove stub

IanKeen commented on Reading List App/Reading List App/FirstViewController.swift:
```diff
@@ -7,19 +7,61 @@
 
+    @IBOutlet weak var welcomeLabel: UILabel!
+    
     override func viewDidLoad() {
```
stub

IanKeen commented:
Perhaps add an ExampleKeys.plist file ? this would make it quicker for devs to get up and running

alexbtlv commented:
Absolutely! I just added ExampleKyes.plist file and removed stub. As for ui/autolayout code, it is a part of [BWWalkthrough framework](https://github.com/ariok/BWWalkthrough/blob/master/BWWalkthrough/BWWalkthroughViewController.swift), without it views are not working properly.

IanKeen commented on Reading List App/Reading List App/AppDelegate.swift:
```diff
@@ -16,6 +18,31 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
+        let twitterConsumerSecret = keys?["twitterConsumerSecret"] as? String
+        
+        // Initialize Parse.
+        Parse.setApplicationId(applicationId!, clientKey: clientKey!)
```
I think this took a step back, you should still do an `if let dict = dict {` check rather than force unwrapping these.
If this is a 'must succeed' section of the app it would be nicer to throw/assert with a descriptive reason rather than confusing people with 'Unexpectedly found nil while unwrapping an optional value'

alexbtlv commented:
@IanKeen could you please take a look at PR, i just commented out all the real log-in/out functionality and made imitation just to play around with log-in screen.

IanKeen commented on Reading List App/Reading List App/AppDelegate.swift:
```diff
@@ -16,6 +18,33 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
+        if let path = NSBundle.mainBundle().pathForResource("Keys", ofType: "plist") {
+            keysDictionary = NSDictionary(contentsOfFile: path)
+        }
+        if let keys = keysDictionary {
```
This if let and the one above can be cleaned up into 1 single statement:
```swift
if let path = NSBundle.mainBundle().pathForResource("Keys", ofType: "plist"),
    let keys = NSDictionary(contentsOfFile: path) {

//use keys dictionary here

}
```
then you can also get rid of the keysDictionary var

IanKeen commented on Reading List App/Reading List App/BWWalkthrough/BWWalkthroughViewController.swift:
```diff
@@ -0,0 +1,286 @@
+    // MARK: - Internal methods -
+    
+    @IBAction func logInButtonPressed() {
+        close(self)
```
this calls the delegate method, but does it also need to call `self.dismissViewControllerAnimated(true, completion: nil)` ? (this seems to be the case when the actual login is successful)

IanKeen commented on Reading List App/Reading List App/FirstViewController.swift:
```diff
@@ -7,19 +7,65 @@
+//       }
+}
 
     override func didReceiveMemoryWarning() {
```
remove stub

IanKeen commented on Reading List App/Reading List App/FirstViewController.swift:
```diff
@@ -7,19 +7,65 @@
+//
+//      Checking if user is actually logged-in
+//
+//      if PFUser.currentUser() != nil {
```
with regards to the live code (when its not commented out) - this is fine here for now, but going forward this shouldn't live in the view controllers (just a note for the future)

fcanas commented:
I do think it's a good time to decide how to manage dependencies. Parse, Bolts, and BWWalkthrough makes for three dependencies in this PR.

pineapple216 commented:
I've also implemented Parse and for now we just use cocoapods. I think we'll be fine with just updating the podfile and not commiting the pods to github. If we clearly specify which version of every library needs to be installed, then we don't run into any problems as far as I'm concerned.
